dashboard/app/controllers/javabuilder_sessions_controller.rb

```diff
@@ -13,40 +13,65 @@ class JavabuilderSessionsController < ApplicationController
 
   # GET /javabuilder/access_token
   def get_access_token
-    teacher_list = get_teacher_list.join(',')
+    unless has_required_params?([:channelId])
+      return render status: :bad_request, json: {}
+    end
     channel_id = params[:channelId]
+
+    session_id = SecureRandom.uuid
+    encoded_payload = get_encoded_payload({session_id: session_id, channel_id: channel_id})
+
+    level_id = params[:levelId].to_i
+    project_files = JavalabFilesHelper.get_project_files(channel_id, level_id)
+    upload_project_files_and_render(session_id, project_files, encoded_payload)
+  end
+
+  # GET /javabuilder/access_token_with_override_sources
+  def get_access_token_with_override_sources
+    unless has_required_params?([:overrideSources])
+      return render status: :bad_request, json: {}
+    end
+    override_sources = params[:overrideSources]
+
+    session_id = SecureRandom.uuid
+    encoded_payload = get_encoded_payload({session_id: session_id})
+
+    level_id = params[:levelId].to_i
+    project_files = JavalabFilesHelper.get_project_files_with_override_sources(override_sources, level_id)
+    upload_project_files_and_render(session_id, project_files, encoded_payload)
+  end
+
+  private
+
+  def upload_project_files_and_render(session_id, project_files, encoded_payload)
+    success = JavalabFilesHelper.upload_project_files(project_files, request.host, encoded_payload)
+    success ?
+      render(json: {token: encoded_payload, session_id: session_id}) :
+      render(status: :internal_server_error, json: {error: "Error uploading sources."})
+  end
+
+  def get_encoded_payload(additional_payload)
+    teacher_list = get_teacher_list.join(',')
     project_version = params[:projectVersion]
     # TODO: remove project_url after javabuilder is deployed with update to no longer need it
     project_url = params[:projectUrl]
     level_id = params[:levelId]
     options = params[:options]
     execution_type = params[:executionType]
-    use_dashboard_sources = params[:useDashboardSources]
+    use_dashboard_sources = 'false'
```

Comment:
sanity check: does this have the correct behavior on the javabuilder side?

Reply:
I just checked that `useDashboardSources` is being correctly set to `false` on the javabuilder side locally. I didn't test against deployed javabuilder, worth doing?

Reply:
I did look up that rails params are always strings, so I feel reasonably confident

Reply:
cool, it's probably ok but since it's not too hard to test against deployed javabuilder now I would test it out to be safe. There can be weirdness with the way things get parsed through the authorizer.

Reply:
😅 well thanks decoupled javabuilder, got a little fast and loose with my refactor and used the wrong session ID key. Updated! Lesson learned to always test deployed javabuilder now that it's so easy.